text/3529-cargo-path-bases.md

```diff
@@ -0,0 +1,395 @@
+`/home/user/dev/rust/libraries/`. Rather than use that path in your
+`Cargo.toml` files, you can define it as a "base" path in
+`~/.cargo/config.toml`:
+
+```toml
+[base-paths]
+dev = "/home/user/dev/rust/libraries/"
+```
+
+Now, you can specify a path dependency on a library `foo` in that
+directory in your `Cargo.toml` using
+
+```toml
+[dependencies]
+foo = { path = "foo", base = "dev" }
+```
+
+Like with other path dependencies, keep in mind that both the base _and_
+the path must exist on any other host where you want to use the same
+`Cargo.toml` to build your project.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Specifying Dependencies
+
+### Base Paths
+
+A `path` dependency may optionally specify a base path by setting the `base` key
+to the name of a base path from the `[base-paths]` table in either the
+[manifest](https://doc.rust-lang.org/cargo/reference/manifest.html) or
+[configuration](https://doc.rust-lang.org/cargo/reference/config.html#base-paths)
+or one of the [built-in base paths](#built-in-base-paths). The value of that
+base path is prepended to the `path` value to produce the actual location where
+Cargo will look for the dependency.
+
+For example, if the Cargo.toml contains:
+
+```toml
+[dependencies]
+foo = { path = "foo", base = "dev" }
+```
+
+Given a `[base-paths]` table in the configuration that contains:
+
+```toml
+[base-paths]
+dev = "/home/user/dev/rust/libraries/"
+```
+
+This will produce a `path` dependency `foo` located at
+`/home/user/dev/rust/libraries/foo`.
+
+If the base path is not found in any `[base-paths]` table or one of the built-in
+base paths then Cargo will generate an error.
+
+If the name of a base path is specified in both the manifest and configuration,
+then the value in the manifest is preferred.
+
+The name of a base path must use only [alphanumeric](https://doc.rust-lang.org/std/primitive.char.html#method.is_alphanumeric)
+characters or `-` or `_`, must start with an [alphabetic](https://doc.rust-lang.org/std/primitive.char.html#method.is_alphabetic)
+character, and must not be empty.
+
+#### Built-in base paths
+
+Cargo provides implicit base paths that can be used without the need to specify
+them in a `[base-paths]` table.
+
+* `workspace` - If a project is [a workspace or workspace member](https://doc.rust-lang.org/cargo/reference/workspaces.html)
+then this base path is defined as the path to the directory containing the root
+Cargo.toml of the workspace.
+
+If one of these built-in base paths is also specified in the manifest or
+configuration, then that value is preferred over the built-in value. This allows
+Cargo to introduce additional built-in paths in the future without compatibility
+issues, as any existing uses will shadow Cargo's built-in name.
+
+## The Manifest Format
+
+[`[base-paths]`](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#base-paths) - Base paths for path dependencies.

```

Comment:
Is this a package or workspace field?

If its a package field, would it support workspace inheritance?  Do workspace dependencies use the inheritable base paths or the package's base paths?

Reply:
I had imagined that this would be a top-level table.

I think supporting workspace inheritance makes sense and that a workspace's dependencies use the workspace's base paths (and an unaware of the member's base paths).

Reply:
Top-level tables are implicitly package tables (e.g. dependencies) or workspace (e.g. patch) or sometimes both.

If we support this as a package field and support workspace inheritance, we'll need to be cautious on the semantics as i worry the interactions between this and dependency inheritance could get too complexd

Reply:
As for 247ff65973d7143ba985ca8481ebf4bad5f6ef1b, that sounds like implicit inheritance and doesn't say how it interacts with dependency inheritance.

Reply:
> doesn't say how it interacts with dependency inheritance

I'm not sure what you're wanting for this - I've clarified how Cargo should look up base paths for members vs workspaces and how relative paths work. Once Cargo has found the base path string and constructed the full path, then the rest of dependency resolution works as per normal.

Reply:
Is the table using workspace inheritance or not?

If you can't, then it should be called out in Rationale and Alternatives as to why we are doing implicit inheritance here, as specified in "Base paths resolution order".

If we are doing workspace inheritance, then we shouldn't mention this in "Base paths resolution order" and instead should call out that users can set
```toml
[base-paths]
workspace = true
```
and the workspace base paths would be made available.  This will require reserving the key name `workspace`.  Like with `[lints]`, as a future possibility, we could allow workspace inheritance *with* overriding of the workspace by the package.

As this is following common `Cargo.toml` practices, only the trade offs of workspace vs package and whether to do workspace inheritance need to be covered in Rationale, and not the design itself (unlike implicit inheritance).

Reply:
See also https://github.com/rust-lang/rfcs/pull/3529#discussion_r1653095733

Reply:
I actually wonder if we shouldn't support this at the workspace-member level and only in the "workspace" root.  Besides making sure we explore and document this, part of my pushing on this is to see what problems crop up from having it at multiple layers.  I think having this only in the root manifest makes dependency inheritance and patch table behavior clearer.  I also see little to gain from having this be used at the package level (granted, I also don't have much in the way of use cases for this outside of the builtin `workspace`).

Reply:
I've removed support for specifying path bases in the manifest so this is moot.